source/m2mtlvserializer.cpp

```diff
@@ -152,87 +155,64 @@ void M2MTLVSerializer::serialize_resource_instance(uint16_t id, M2MResourceInsta
-    memset(tlv,0,size + type_length + id_size + length_size + value_length+1);
+    tlv = (uint8_t*)malloc(size + type_length + id_size + length_size + value_length);
     if(data) {
         memcpy(tlv, data, size);
+        free(data);
     }
-    memcpy(tlv+size, tlv_type, type_length);
-    memcpy(tlv+size+type_length, id_ptr, id_size);
-    memcpy(tlv+size+type_length+id_size, length_ptr, length_size);
+    memcpy(tlv+size, &tlv_type, type_length);
+    memcpy(tlv+size+type_length, id_array, id_size);
+    memcpy(tlv+size+type_length+id_size, length_array, length_size);
     memcpy(tlv+size+type_length+id_size+length_size, value, value_length);
 
-    free(tlv_type) ;
-    free(length_ptr);
-    free(id_ptr);
-    free(data);
-
     data = tlv;
     size += type_length + id_size + length_size + value_length;
 }
 
-uint8_t* M2MTLVSerializer::serialize_id(uint16_t id, uint32_t &size)
+void M2MTLVSerializer::serialize_id(uint16_t id, uint32_t &size, uint8_t *id_ptr)
 {
-    uint32_t id_size = id > 255 ? 2 : 1;
-    uint8_t *id_ptr = (uint8_t*)malloc(id_size);
-    memset(id_ptr, 0 , id_size);
-    size += id_size;
     if(id > 255) {
-        *id_ptr = (id & 0xFF00) >> 8;
-        id_ptr++;
-        *id_ptr = id & 0xFF;
-        id_ptr--;
+        size=2;
+        id_ptr[0] = (id & 0xFF00) >> 8;
+        id_ptr[1] = id & 0xFF;
     } else {
-        *id_ptr = id & 0xFF;
+        size=1;
+        id_ptr[0] = id & 0xFF;
     }
-    return id_ptr;
 }
 
-uint8_t* M2MTLVSerializer::serialize_length(uint32_t length, uint32_t &size)
+void M2MTLVSerializer::serialize_length(uint32_t length, uint32_t &size, uint8_t *length_ptr)
 {
-    uint8_t *length_ptr = 0;
-    uint32_t length_size = 0;
     if (length > 65535) {
-        length_size = 3;
-        length_ptr = (uint8_t*)malloc(length_size+1);
-        memset(length_ptr,0,length_size+1);
-        *length_ptr = (length & 0xFF0000) >> 16;
-        length_ptr++;
-        *length_ptr = (length & 0xFF00) >> 8;
-        length_ptr++;
-        *length_ptr = length & 0xFF;
-        length_ptr--;
-        length_ptr--;
+        size = 3;
+        length_ptr[0] = (length & 0xFF0000) >> 16;
+        length_ptr[1] = (length & 0xFF00) >> 8;
+        length_ptr[2] = length & 0xFF;
     } else if (length > 255) {
-        length_size = 2;
-        length_ptr = (uint8_t*)malloc(length_size+1);
-        memset(length_ptr,0,length_size+1);
-        *length_ptr = (length & 0xFF00) >> 8;
-        length_ptr++;
-        *length_ptr = length & 0xFF;
-        length_ptr--;
+        size = 2;
+        length_ptr[0] = (length & 0xFF00) >> 8;
+        length_ptr[1] = length & 0xFF;
     } else if (length > 7) {
-        length_size = 1;
-        length_ptr = (uint8_t*)malloc(length_size+1);
-        memset(length_ptr,0,length_size+1);
-        *length_ptr = length & 0xFF;
+        size = 1;
+        length_ptr[0] = length & 0xFF;
+    } else {
+        size=0;
     }
-    size += length_size;
```

Comment:
Original code added the length to given parameter value. Wasn't this feature used by caller, or is it broken?